File: backend/src/main/java/com/btt/continew/contract/domain/Contract.java
package com.btt.continew.contract.domain;

import com.btt.continew.contract.controller.dto.request.ContractRequest;
import com.btt.continew.global.domain.BaseEntity;
import com.btt.continew.house.domain.House;
import com.btt.continew.member.domain.Member;
import java.time.LocalDate;
import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.FetchType;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.JoinColumn;
import javax.persistence.ManyToOne;
import lombok.Builder;
import lombok.Getter;
import org.hibernate.annotations.Where;

@Getter
@Entity
@Where(clause = "deleted_at is null")
public class Contract extends BaseEntity {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "contract_id")
    private Long id;

    @ManyToOne(targetEntity = House.class, fetch = FetchType.LAZY)
    @JoinColumn(name = "house_id", nullable = false)
    private House house;

    @ManyToOne(targetEntity = Member.class, fetch = FetchType.LAZY)
    @JoinColumn(name = "seller_id", nullable = false)
    private Member seller;

    @ManyToOne(targetEntity = Member.class, fetch = FetchType.LAZY)
    @JoinColumn(name = "buyer_id", nullable = false)
    private Member buyer;

    @Column(name = "level")
    private Integer level;

    @Column(name = "location")
    private String location;

    @Column(name = "area")
    private String area;

    @Column(name = "net_leasable_area")
    private String netLeasableArea;

    @Column(name = "contract_type")
    private String contractType;

    @Column(name = "tenancy_deposit")
    private String tenancyDeposit;

    @Column(name = "maintenance_fee")
    private String maintenanceFee;

    @Column(name = "contract_start")
    private LocalDate contractStart;

    @Column(name = "contract_end")
    private LocalDate contractEnd;

    @Column(name = "total_premium")
    private String totalPremium;

    @Column(name = "down_payment")
    private String downPayment;

    @Column(name = "middle_payment")
    private String middlePayment;

    @Column(name = "middle_date")
    private LocalDate middleDate;

    @Column(name = "balance_payment")
    private String balancePayment;

    @Column(name = "balance_date")
    private LocalDate balanceDate;

    @Column(name = "seller_address")
    private String sellerAddress;

    @Column(name = "seller_name")
    private String sellerName;

    @Column(name = "seller_birth")
    private LocalDate sellerBirth;

    @Column(name = "seller_phone", length = 16)
    private String sellerPhone;

    @Column(name = "seller_signature")
    private String sellerSignature;

    @Column(name = "buyer_address")
    private String buyerAddress;

    @Column(name = "buyer_name")
    private String buyerName;

    @Column(name = "buyer_birth")
    private LocalDate buyerBirth;

    @Column(name = "buyer_phone", length = 16)
    private String buyerPhone;

    @Column(name = "buyer_signature")
    private String buyerSignature;

    public Contract() {
    }

    @Builder
    public Contract(House house, Member seller, Member buyer) {
        this.house = house;
        this.seller = seller;
        this.buyer = buyer;
        this.level = 1;
    }

    public void levelOneWrite(ContractRequest request) {
        this.location = request.getLocation();
        this.area = request.getArea();
        this.netLeasableArea = request.getNetLeasableArea();
        this.contractType = request.getContractType();
        this.tenancyDeposit = request.getTenancyDeposit();
        this.maintenanceFee = request.getMaintenanceFee();
        this.contractStart = request.getContractStart();
        this.contractEnd = request.getContractEnd();
        this.totalPremium = request.getTotalPremium();
        this.downPayment = request.getDownPayment();
        this.middlePayment = request.getMiddlePayment();
        this.middleDate = request.getMiddleDate();
        this.balancePayment = request.getBalancePayment();
        this.balanceDate = request.getBalanceDate();
        this.sellerAddress = request.getSellerAddress();
        this.sellerName = request.getSellerName();
        this.sellerBirth = request.getSellerBirth();
        this.sellerPhone = request.getSellerPhone();
        if (request.getNextLevel()) {
            this.level = 2;
        }
    }

    public void levelTwoWrite() {

    }

    public void levelThreeWrite() {

    }
}

File: frontend/src/store/contract.ts
import { ContractType } from "src/types/contractType";

const CONTRACT = "user/setContractInfo";
const ID = "user/setContractIds";
const LEVEL = "user/setContractLevel";
const ROLE = "user/setRole";
const STEP1 = "user/step1";
const STEP2 = "user/step2";
const STEP3 = "user/step3";

interface actionType {
	type: string;
	data: ContractType;
}

interface required {
	buyer: string;
	house_id: number;
	seller: string;
}

export const SET_CONTRACT = (data: ContractType) => ({ type: CONTRACT, data });
export const SET_ID = (data: required) => ({ type: ID, data });
export const SET_LEVEL = (data: boolean) => ({ type: LEVEL, data });
export const SET_ROLE = (data: string) => ({ type: ROLE, data });
export const SET_STEP1 = (data: ContractType) => ({ type: STEP1, data });
export const SET_STEP2 = (data: string) => ({ type: STEP2, data });
export const SET_STEP3 = (data: string) => ({ type: STEP3, data });

const initialState = {
	contract: {
		location: "",
		area: 0,
		net_leasable_area: 0,
		contract_type: "",
		tenancy_deposit: "",
		maintenance_fee: "",
		contract_start: "",
		contract_end: "",
		total_premium: "",
		down_payment: "",
		middle_payment: "",
		middle_date: "",
		balance_payment: "",
		balance_date: "",
		seller_address: "",
		seller_name: "",
		seller_birth: "",
		seller_phone: "",
		seller_signature: "",
		buyer_address: "",
		buyer_name: "",
		buyer_birth: "",
		buyer_phone: "",
		buyer_signature: "",
	},
	level: {
		next_level: true,
	},
	id: {
		house_id: 0,
		seller_login_id: "",
		buyer_login_id: "",
	},
	step: {
		current_step: 0,
	},
	role: {
		user_role: "",
	},
};

function contractInfo(state = initialState, action: actionType) {
	switch (action.type) {
		case CONTRACT:
			return {
				...state,
				contract: {
					...state.contract,
					location: action.data.location,
					area: action.data.area,
					net_leasable_area: action.data.net_leasable_area,
					contract_type: action.data.contract_type,
					tenancy_deposit: action.data.tenancy_deposit,
					maintenance_fee: action.data.maintenance_fee,
					contract_start: action.data.contract_start,
					contract_end: action.data.contract_end,
					total_premium: action.data.total_premium,
					down_payment: action.data.down_payment,
					middle_payment: action.data.middle_payment,
					middle_date: action.data.middle_date,
					balance_payment: action.data.balance_payment,
					balance_date: action.data.balance_date,
					seller_address: action.data.seller_address,
					seller_name: action.data.seller_name,
					seller_birth: action.data.seller_birth,
					seller_phone: action.data.seller_phone,
					seller_signature: action.data.seller_signature,
					buyer_address: action.data.buyer_address,
					buyer_name: action.data.buyer_name,
					buyer_birth: action.data.buyer_birth,
					buyer_phone: action.data.buyer_phone,
					buyer_signature: action.data.buyer_signature,
				},
				step: { ...state.step, current_step: action.data.current_level },
			};
		case ID:
			return {
				...state,
				id: {
					...state.id,
					house_id: action.data.house_id,
					seller_login_id: action.data.seller,
					buyer_login_id: action.data.buyer,
				},
			};
		case LEVEL:
			return {
				...state,
				level: {
					next_level: action.data.next_level,
				},
			};
		case ROLE:
			return {
				...state,
				role: {
					user_role: action.data,
				},
			};
		case STEP1:
			return {
				...state,
				contract: {
					...state.contract,
					location: action.data.location,
					area: action.data.area,
					net_leasable_area: action.data.net_leasable_area,
					contract_type: action.data.contract_type,
					tenancy_deposit: action.data.tenancy_deposit,
					maintenance_fee: action.data.maintenance_fee,
					contract_start: action.data.contract_start,
					contract_end: action.data.contract_end,
					total_premium: action.data.total_premium,
					down_payment: action.data.down_payment,
					middle_payment: action.data.middle_payment,
					middle_date: action.data.middle_date,
					balance_payment: action.data.balance_payment,
					balance_date: action.data.balance_date,
					seller_address: action.data.seller_address,
					seller_name: action.data.seller_name,
					seller_birth: action.data.seller_birth,
					seller_phone: action.data.seller_phone,
				},
			};
		case STEP2:
			return {
				...state,
				contract: {
					...state.contract,
					buyer_signature: action.data,
				},
			};
		case STEP3:
			return {
				...state,
				contract: {
					...state.contract,
					seller_signature: action.data,
				},
			};
		default:
			return state;
	}
}

export default contractInfo;
